icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java

```diff
@@ -1239,147 +1563,54 @@ public static String canonicalize(String localeID){
-                            (new StringBuilder(parser.getLanguage().length() + 1 + parser.getVariant().length()))
-                                .append(parser.getLanguage())
-                                .append("_")
-                                .append(variant)
-                                .toString())
-                            .get("replacement")
-                            .getString());
-                        StringBuilder replacedVariant = new StringBuilder(parser.getVariant().length());
-                        for (String current : variants) {
-                            if (current.equals(variant)) continue;
-                            if (replacedVariant.length() > 0) replacedVariant.append("_");
-                            replacedVariant.append(current);
-                        }
-                        parser = new LocaleIDParser(
-                            (new StringBuilder(localeID.length()))
-                                .append(lscvToID(replaceLocale.getLanguage(),
-                                    !parser.getScript().isEmpty() ? parser.getScript() : replaceLocale.getScript(),
-                                    !parser.getCountry().isEmpty() ? parser.getCountry() : replaceLocale.getCountry(),
-                                    replacedVariant.toString()))
-                                .append(parser.getName().substring(parser.getBaseName().length()))
-                                .toString());
-                    } catch (MissingResourceException e) {
-                    }
-                }
-            }
-
-            // language _ script _ country
-            // ug_Arab_CN -> ug_CN
-            if (!parser.getScript().isEmpty() && !parser.getCountry().isEmpty()) {
-                try {
-                    ULocale replaceLocale = new ULocale(languageAlias.get(
-                        lscvToID(parser.getLanguage(), parser.getScript(), parser.getCountry(), null))
-                        .get("replacement")
-                        .getString());
-                    parser = new LocaleIDParser((new StringBuilder(localeID.length()))
-                        .append(lscvToID(replaceLocale.getLanguage(),
-                            replaceLocale.getScript(),
-                            replaceLocale.getCountry(),
-                            parser.getVariant()))
-                        .append(parser.getName().substring(parser.getBaseName().length()))
-                        .toString());
-                } catch (MissingResourceException e) {
-                }
+        boolean knownCanonicalized = false;
+        String name = parser.getName();
+        if (!isKnownCanonicalizedLocale(name)) {
+            AliasReplacer replacer = new AliasReplacer(
+                parser.getLanguage(), parser.getScript(), parser.getCountry(), parser.getVariant().toLowerCase(),
+                parser.getName().substring(parser.getBaseName().length()));
```

Comment:
getName() is in ICU syntax, so if there are extensions, this will start with `@`. intentional? harmless? comment? tested?

Reply:
parser.getName().substring(parser.getBaseName().length()) 
would get the part of of the locale starting "@" to the AliasReplacer, and the the replace() method will append them back. 
There are test cases cover that already